Intro/README.md

```diff
@@ -0,0 +1,455 @@
+# Introduction to TPMs
+
+Trusted Platform Modules (TPMs) are a large and complex topic, made all
+the more difficult to explain by the intricate relationships between the
+relevant concepts.  This is an attempt at a simple explanation --
+simpler than reading hundreds of pages of documents, but then too, too
+light on detail to be immediately useful.
+
+So what is a TPM?  Well, it's a cryptographic co-processor with special
+features to enable "root of trust measurement" (RTM), remote attestation
+of system state, unlocking of local resources that are kept encrypted
+(e.g., filesystems), and more.  A TPM can do those things, and it can do
+it with rich authentication and authorization policies.
+
+Typically a TPM is a hardware module, a chip, though there are firmware,
+virtual, and simulated TPMs as well, all implemented in software.
+
+To simplify things we'll consider only TPM 2.0.  Also to simplify things
+we'll ignore algorithm agility.
+
+Other parts of this [tutorial](README.md) may cover specific concepts in
+much more detail.
+
+# Core Concepts
+
+Some core concepts in the world of TPMs
+
+ - cryptography:
+    - digest functions
+    - hash extension
+    - HMAC
+    - symmetric encryption
+    - asymmetric encryption and/or signature algorithms
+ - platform configuration registers (PCRs)
+ - cryptographic object naming
+ - immutability of object public areas
+ - key hierarchies
+ - key wrapping
+ - limited resources
+    - tickets
+    - resource management
+ - sessions
+ - authorization
+    - restricted cryptographic keys
+    - policies
+ - other object types
+    - non-volatile (NV) indexes
+ - attestation
+
+We'll assume reader familiarity with cryptography so we need not explain
+it.
+
+Authorization is the most important aspect of a TPM, since that's
+ultimately what it exists for: to authorize a system or application to
+perform certain duties when all the desired conditions allow for it.
+
+TPMs have a very rich set of options for authorization.  It's not just
+[policies](#Policies), but also cryptographic object names used with
+restricted keys to allow access only to applications that also have
+other access.
+
+Where to start?  Let's start with hash extension, which may be the only
+trivial concept in the world of TPMs!
+
```

Comment:
I think this needs to be restructued - I'm already lost here.  To introduce authorisations we need to introduce hierarchies first and before that what a TPM has inside, ie:  Objects and Hierarchies..   Objects can be keys, NVRAM slots, PCRs etc,  Hierarchies - we have 4 of though .

Reply:
I kinda disagree though.  I think EA policies are a perfect example of the power of hash extension, and extremely clever.  EA policies aren't all that interesting as an introductory matter, but their cleverness is, and, IMO, it is authorization in all its flavors that makes TPMs so useful.  I want to give the reader a flavor for the flexibility and power of TPMs, but I want to do it in just a few pages, not hundreds.

Reply:
I've read "A Practical Guide to TPM 2.0", and I highly, highly recommend that book.  It worked for me.  

Because all these TPM topics are so interrelated you get the old `B before A` pattern.  Want to learn quantum mechanics?  You're going to have to learn... so much other stuff before you start on that.  But for many students that approach is just discouraging -- for them "let's study just enough of B to get started on A, then go back to B later" works very well.

In the case of TPMs there's just... so much of of the `B before A` pattern, and even some circularity...  You really see a fair bit of this in "A Practical Guide to TPM 2.0", where many chapters cover some related topic lightly and then later chapters cover that ancillary topic in depth.

I don't mind the `B- before A then A then B+ then A+` pattern, but I need a more accessible document for people who mostly don't need to become experts in TPMs, and hopefully such a document can help _some_ of its audience get started on becoming experts.

One has to start somewhere.  Where?  I think hash extension is the obvious starting place because it is core to PCRs and policies.  Also cryptographic object naming: because it too is central to authorization!  And authorization is, IMO, _the_ essential feature of TPMs, so it is my goal to get to it, and I want to get to it very quickly because I don't want to bore and drive away the lay reader.

Obviously explaining EA policies in all their detail requires first explaining everything else.  Yet since policies are so interesting, leaving them for last is a bit disappointing.  My approach right now is: explain a little bit of each concept, then explain policies, then stop.

My take is that key hierarchies are not as essential to understanding policies as you seem to think.  Maybe... maybe that's wrong.  I don't know.  Given my goal here I need to hear from reviewers who literally know nothing about TPMs.  I also want review by TPM experts, both as to errors in my text, and as to opinion on how to structure this, but while I will consider such opinions and maybe accept them, I might also reject some opinions.

Eventually I -or someone else- might write in-depth coverage of each topic, but still having this very light umbrella treatment should help approach each topic's in-depth coverage.

Reply:
TPM 2.0 Authorization and Extended Authorization(EA) is one of the most complex topics. 

I think I would read through the tutorial once again after few days, so I have fresh eyes and then will give feedback on this paragraph.

Reply:
I've done some restructuring.

Reply:
> TPM 2.0 Authorization and Extended Authorization(EA) is one of the most complex topics.

But I don't cover it in detail.  And I think it's especially illustrative of the brilliance of using digest values to identify things that don't fit in a TPM's resources.  PCRs and policies both use hash extension in this way.  One needs an externally-maintained eventlog to reconstruct how a PCR got to some value.  One needs to know a policy in order to send it to the TPM for evaluation, then at the end the TPM checks that its digest _is_ the policy associated with whatever resource is being accessed.

I might further restructure the intro to make this particular idea more prominent but cover policies in less detail.

Reply:
Note that, as complex as EA policies are, I don't have that much text on the topic.

Reply:
".. which may be the only trivial concept in the world of TPMs!"  Too much opinion for my taste.  There's a lot of "this is really hard" throughout.

Reply:
Ok.